Enable automating static elements workflow

sdrenth opened the pull request:
### What does it do?
Added settings to allow for an automated static elements workflow and added the handling of creating/removing automated static files. 

**Features:**
-  Added System settings allowing an automated static elements workflow for each element type (Templates, Template Variables, Chunks, Snippets and Plugins)
- Added System setting for setting a default mediasource for your static elements
- Added System setting for setting a default category for your static elements
- Added System setting for setting a static elements basepath
- If set, will automatically create a static file where the static filename is based on the name of the element, stripping it from special characters, removing spaces and making it lowercased.
- When removing/deleting a static element, it will remove the static file (if not used by another element) and will also remove the parent directory automatically for you if the directory is empty.
- Added static file path to duplicate element window

### Why is it needed?
We use static elements in our workflow and by automating the creation of static elements it will save a lot of manual work and it also keeps the static elements free of clutter by removing unused static element files.

sdrenth committed:
Automating static elements

sdrenth committed:
Automating static elements WIP

sdrenth committed:
Automating static elements WIP

sdrenth committed:
Dont remove template if static file is still being used by other template

sdrenth committed:
Finishing automating static elements

sdrenth committed:
Code improvements and fixes

ComSeo commented:
So cool... looking forward to this!

sdrenth committed:
Move static file after drag and dropping an element

gpsietzema commented:
+1

Mark-H commented:
This looks like it must've been a fair bit of work! Kudos @sdrenth :) I'm getting a little too tired to review this one during today's merge session, but a first glance looks fairly promising. It would be nice to have a bit more information in terms of "user docs" for how people could use that, where files go, etc, so we can add that straight into the docs when it's merged.

sdrenth commented:
Yes it was @Mark-H but it also makes using static elements a lot quicker! :) Here is some additional documentation.

## Automating your static elements workflow
It is now possible to automate the process of generating static elements. It takes away the manual steps of creating a static element such as marking an element as static, selecting your static elements media source and it automatically generates the filepath name for your static file.

### Key features
- Automates marking an element as static, selecting the default media source, selecting a default category (optional) and generates a filename based on a specified basepath, element type and filename.
- Removes old static files when a static element has been removed or renamed

### How to get setup
1. Go to system settings: Select Core as namespace and Static Elements as topic.
2. Turn on automating static elements for each element type you want to use automate the static elements workflow for (For example snippets and chunks).
3. Select your default media source to use for static elements.
4. Provide your static elements basepath, for example: components/static/elements/.
5. Optionally select a default category to use for your static elements.

### File structure
Files are stored as: {static_elements_basepath}/{type}/{category}/{filename}.{extension} and are stored within the specified media source. For example: components/site/elements/chunks/html/footer.chunk.tpl.

bezumkin commented:
> We use static elements in our workflow and by automating the creation of static elements it will save a lot of manual work and it also keeps the static elements free of clutter by removing unused static element files.

But you still need to create every single element in manager. You can`t just upload files from IDE and use it in your code right after this.

sottwell commented:
I think this should all be an extra, much like a gallery extra. The general idea is to reduce the core, not add features that the vast majority of users will never use.

Oetzie commented:
@sdrenth Can you also fix the mediasource browser bug wit this pull request? When you update a chunk, plugin or another element the selected 'mediasource' is not set correctly to the 'static_file' field. When you change the 'mediasource' field, an listener will be trigger that will set the 'mediasource' for the 'static_file' field. So every time when you want to change the file you need to select a 'mediasource' first.

To fix this add the following line to the element panels (chunk panel, plugin panel, template panel etc):
```,config.record.source != null ? config.record.source : MODx.config.default_media_source```
Just add the line before the 'openTo' config from the 'static_file' field.

Oetzie commented:
@sottwell 'Static elements' is an feature in MODX, but it is just a half working feature. With this pull request it can improved so much... 'Static elements' is now just a half working feature so i would love this pull request. Don't make an extra for this, this is just a must have for 'static elements'....

sdrenth commented:
@Oetzie Since that bug is not related to this PR I think it is better if you'd create a new PR where you add this fix into it. 

@sottwell I agree with Oetzie that this PR should be part of the MODX core and should not be an extra, since static elements is already part of the core.

sonicpunk commented:
I would like to test this at the #modxbughunt

sonicpunk commented:
@sdrenth in my testing I came upon a strange behaviour when you change the category of an already registered chunk. In the file system the chunk file just disappears.

How I reproduced this:

- created a chunk
- assigned a category
- changed the category

Result is that the file gets deleted from the filesystem. The expected behaviour would be that a new folder named after the category is created and in this newly created folder you would find the chunk.

<img width="1247" alt="screen shot 2018-02-09 at 12 01 03" src="https://user-images.githubusercontent.com/447453/36024847-4fd93fe6-0d91-11e8-90db-1ff0151254c2.png">

I love the ideas here and would love to see a solution. It being a core feature would make MODX out-of-the-box more attractive to developers. 

What @bezumkin says I also agree with. The preferred workflow would be that you only work in the IDE and the files get automatically created in MODX based on maybe the file structure?

sdrenth commented:
@sonicpunk It should remove the old static file and store the file in a new path including the new category name. I just tested it the following works fine with me:

* Created chunk and added some content
* Assigned to category: Html
* Save: I see that the static file is created: /chunks/html/test.chunk.tpl
* Changed category to: Slider
* Save: I see the old static file is removed and static file is created: /chunks/slider/test.chunk.tpl

Can you provide the exact steps that I can follow to reproduce this issue?

Regarding @bezumkin his comment, I see his point, but for now I think this PR greatly improves the static elements workflow. Perhaps he can create a new issue for this so that can be picked up separately? This PR is more focused on the improving the default MODX static elements workflow, not refactoring the static elements workflow.

sonicpunk commented:
@sdrenth I tested a little more and seem to be able reproduce this behaviour when I do the following:

1) Create Chunk - no category attached
2) Create new category
3) Assign category to the chunk

The chunk gets deleted and the category is not written to the chunks folder.
Maybe it is the order that I am doing this.

superspace commented:
@sonicpunk @sdrenth I can reproduce the behaviour sonicpunk ist reporting – but it only happens, if the chunk has no content. With some html inside, it behaves as sdrenth is describing.
#modxbughunt

sdrenth commented:
@sonicpunk @superspace Yeah, that way i can also reproduce the issue as described by sonicpunc. So it's not actually an issue I suppose? Or is there a reason why it then should create an empty file?

superspace commented:
@sdrenth If your intention is to edit the element in your IDE it makes sense to create an empty file.

sonicpunk commented:
@sdrenth Yes, my use case would be just to set it all up with empty files and then edit in the IDE.

sdrenth committed:
Fix to create empty static files for elements

sdrenth committed:
Fix if check

sdrenth commented:
@sonicpunk @superspace I've added a check so empty static files are now created. Although I cant seem to get the new commit to pass the Travis CI checks? 

Can someone explain to me why Travis CI is failing?

Mark-H commented:
If you click on "X" by the commit you can see the task on travis, and drill down into the failures for the different versions. Looks like a whole bunch of the unit tests are failing.

sdrenth committed:
Fixing saving empty static files

sdrenth committed:
Fixing saving empty static files

sdrenth commented:
@Mark-H It passed the CI tests now ;)

gpsietzema commented:
Time to integrate then 👍

sonicpunk commented:
Looking forward to this!

christianseel commented:
Can we pretty please get this merged for MODX 3?

alroniks commented:
@christianseel I'm going to check it tomorrow and fix conflicts that appear here but it still requires approvals from code owners. Kick Mark :)

Mark-H commented:
Hugs are more effective than kicks!

alroniks commented:
It's a joke, @Mark-H. I promise we will hug at the next meet. :)

thewhiterabbit commented:
Pulling this to test/resolve any conflicts I can resolve. Much needed feature. Thank you @sdrenth!

thewhiterabbit commented:
PS: I will resolve any conflicts as I did before with the mySQL upgrade issues. I'd love to help any way I can. Sorry I've been away for two years.

thewhiterabbit commented:
I'm currently testing this and so far I love it!

thewhiterabbit commented:
I did a merge on this with origin/3.x branch with @OptimusCrime's help, and I have made one resolution so far. If this looks correct I will proceed with the other conflict resolutions... 
![b5ef26ad-0e52-4d71-9446-5f0d321219f5](https://user-images.githubusercontent.com/5334222/45905901-ff491300-bda6-11e8-9263-62d5311ef5c7.jpeg)
![07f6708b-f0b8-404e-a71c-3effe2b1f298](https://user-images.githubusercontent.com/5334222/45905902-ff491300-bda6-11e8-89c3-b5b6dd462378.jpeg)

Please advise. 🤓

OptimusCrime commented:
Hey man :) Would it not make more sense to wait for this to be merged into 2.x, and then for the entire 2.x branch to be merged into 3.x? Is there any reason why this needs to be added into 3.x before all the bugfixes ect are carried over?

thewhiterabbit commented:
> Hey man :) Would it not make more sense to wait for this to be merged into 2.x, and then for the entire 2.x branch to be merged into 3.x? Is there any reason why this needs to be added into 3.x before all the bugfixes ect are carried over?

Not really any reason, but since it was requested to pull into 3.x,

> Can we pretty please get this merged for MODX 3?

I wanted to test it there also. I already successfully tested it in 2.x branch local running php 7.

Update:
Tested in 3.x and it is not creating the files. Checked the write access and it's correct. Further investigation shows that the JS files I resolved are working, as the field does change with the lower case file name as expected. Will continue to update this as I go.

Moving back to the 2.x branch to test further, then moving up the version line. I have a feeling that you're very correct in that statement above @OptimusCrime. Will try to merge working 2.x branch I've built on local into the 3.x branch after further testing verifies no errors in each successive branch.

Update:
Tested up to 2.6.x branch working! No resolutions were needed for merge with 2.6x branch.

gpsietzema commented:
I tested this thoroughly. Results:
- Creating elements works as expected
- Moving them around works perfectly
-

gpsietzema commented:
v Creating elements works as expected
v Editing elements works as expected
v Moving elements works as expected
v The issue Ben mentioned was also fixed

x Moving categories doesn't do anything, but thinking more about it I think it is the right behaviour. It would become too complex.

So it can be merged if you ask me :-)

opengeek commented on core/xpdo/om/xpdoobject.class.php:
```diff
@@ -1122,6 +1122,7 @@ public function & getOne($alias, $criteria= null, $cacheFlag= true) {
                 }
             }
+
             if ($object= $this->xpdo->getObject($fkdef['class'], $criteria, $cacheFlag)) {
```
The modification to this file needs to be removed.

opengeek commented on core/model/modx/modprocessor.class.php:
```diff
@@ -947,6 +947,7 @@ class modObjectDuplicateProcessor extends modObjectProcessor {
     public $newObject;
     public $nameField = 'name';
+    public $staticfileField = 'static_file';
     /** @var string $newNameField The name of field that used for filling new name of object.
```
Shouldn't the changes being made here to the modObjectDuplicateProcessor be moved to a dedicated modStaticElementDuplicateProcessor? This seems like it could affect any object being duplicated.

sdrenth replied:
@opengeek It won't do anything if the object that is being duplicated is not static so I dont think this is really necessary

sdrenth committed:
Removing newline

opengeek commented:
@sdrenth I think I am out of my element attempting to resolve the current conflicts on this PR. Is there any chance you could take a stab at rebasing this work on top of the latest 2.x branch?

sdrenth commented:
@opengeek I've added a new PR for the rebased version:
https://github.com/modxcms/revolution/pull/14135
